File: feb2/mar15_Rec12/wordbreak.java
package mar15;
import java.util.*;
public class wordbreak {

	public static boolean isTheWordInVocab(ArrayList<String> vocab,String word){
		for(int i=0;i<vocab.size();i++){
			if(vocab.get(i).compareTo(word)==0)
				return true;
		}
		return false;
	}
	public static void mango(ArrayList<String> vocab,String q,String a){
		if(q.length()==0){
			System.out.println(a);
			return;
		}
		
		for(int d=0;d<q.length();d++){
			String s=q.substring(0, d+1);
			String roq=q.substring(d+1);
			if(isTheWordInVocab(vocab, s))
				mango(vocab,roq,a+s+"-");
		}
	}
	
	public static void main(String[] args) {
		ArrayList<String> vocab=new ArrayList<>();
		vocab.add("man"); vocab.add("go"); vocab.add("mango"); vocab.add("ice"); vocab.add("cream"); vocab.add("icecream"); vocab.add("air");
		vocab.add("plane"); vocab.add("airplane");
		mango(vocab,"gomanicecreamairplane","");
	}

}

File: june14/july13_Rec9/wordbreak.cpp
#include<iostream>
#include<string>
#include<unordered_set>

using namespace std;

// bool isthewordinvocab(unordered_set<string>& vocab,string word)
// {
//     if(vocab.count(word) == 1)
//         return true;
//     return false;
// }
void mango(unordered_set<string>& vocab,string query,string ans)
{
    if(query.length() == 0)
    {
        cout << ans << endl;
        return;
    }
    for(int i=0;i<query.length();i++)
    {
        string qs=query.substr(0,i+1);
        string ros=query.substr(i+1);
        if(vocab.count(qs) == 1)
        {
            mango(vocab,ros,ans+qs+" ");
        }
    }
}
int main(int argc,char** argv)
{
    // vector<string> dictionary {"man","go","mango","ice","cream","icecream","air","plane","airplane"};
    // vector<string> dictionary {"like","sung","go","man","mango","sam","samsung","i"};
    // mango(dictionary,"gomanicecreamairplane","");
    // mango(dictionary,"ilikesamsungmango","");
    unordered_set<string> dictionary;
    dictionary.insert("man");
    dictionary.insert("go");
    dictionary.insert("mango");
    dictionary.insert("ice");
    dictionary.insert("cream");
    dictionary.insert("icecream");
    dictionary.insert("sam");
    dictionary.insert("sung");
    dictionary.insert("samsung");
    mango(dictionary,"mangoicecreamsamsung","");

}